homeassistant/components/zha/binary_sensor.py

```diff
@@ -48,41 +49,15 @@
 
 async def async_setup_entry(hass, config_entry, async_add_entities):
     """Set up the Zigbee Home Automation binary sensor from config entry."""
-
-    async def async_discover(discovery_info):
-        await _async_setup_entities(
-            hass, config_entry, async_add_entities, [discovery_info]
-        )
+    entities = hass.data[DATA_ZHA][DOMAIN] = []
```

Comment:
Where is this populated?

Should we rename this too?

Reply:
it's populated here https://github.com/Adminiuga/home-assistant/blob/9b9e23a79e7c0bcda2e676f97a5e483c97f96b7b/homeassistant/components/zha/core/channels/__init__.py#L123

I'm open for naming suggestions, since it is a list of entity's class and class initialization params. 

On a side note, are we allowed storing `async_add_entities` instead? if we are, then potentially we could just store/register `async_add_entities` with zha.core.discovery class when setting component and calling it later directly from zha.core.discovery instance

Reply:
I think we are. I am pretty sure I saw balloob suggest to someone else to do exactly that (store a ref to the async_add_entities method) earlier this week somewhere...

Reply:
![image](https://user-images.githubusercontent.com/1335687/75043626-654f3380-548e-11ea-85d7-da1322f89ad5.png)

Reply:
I need to take a second look to ZhaEntity tie to the `ZhaDevice`. Could cleanuo a few things if we could remove it

Reply:
I think we could untie ZhaDevice from ZhaEntity but I'd rather leave it for another PR.

Reply:
so will anything be done here with regard to `async_add_entities` on this PR?

Reply:
It's ok to store `async_add_entities` but it should only be called from the platform. It's ok to eg extract a common function that is defined in a central place, that calls `async_add_entities`, but then this function must be called from within the corresponding platform. Creating and adding entities are a platform concern and must be done within platforms.

Reply:
Names are hard.

`entity_info`
`entity_cls_info`
`entities_for_creation`

Nothing really sticks.

Reply:
Was thinking about `pre_entities` :) but decided to go with `entities_to_create`.

Reply:
> It's ok to eg extract a common function that is defined in a central place, that calls async_add_entities, but then this function must be called from within the corresponding platform

So this is where I'm lost: is attaching that "common" function to a "signal" in corresponding platform counts? even if the signal is not originating in the corresponding platform but in ZHA integration?

Reply:
Attaching a signal in the platform is ok. The signal doesn't know what the target it. The platform knows what the target is.

Reply:
Sounds good. Let's keep it as is for now.